src/Guides/RestructuredText/Parser/DocumentParser.php

```diff
@@ -200,12 +200,17 @@ private function parseLines(string $document): void
         $this->documentIterator->load($this->environment, $document);
         $this->setState(State::BEGIN, '');
 
-        foreach ($this->documentIterator as $line) {
+        // We explicitly do not use foreach, but rather the cursors of the DocumentIterator
+        // this is done because we are transitioning to a method where a Substate can take the current
+        // cursor as starting point and loop through the cursor
+        while ($this->documentIterator->valid()) {
```

Comment:
I like this :+1:  it makes sense to me that each Production has control over when it stops parsing.

Reply:
🎉